File: clients/defold-app/playground/labs/lab11.lua
--
-- Lab 11 — Deterministic Randomness.
--
-- The shotgun fan is "random", yet the client predicts every pellet EXACTLY —
-- because the randomness is a pure function of data both sides already share:
--
--   seed = splitmix32(input seq ^ room salt)  ->  mulberry32 pellet stream
--
-- The seq is the engine's own input counter (input:send() returns it; the server
-- reads channel.consumedCount). The salt is synced room state. NOTHING about the
-- pellets rides the wire — the server's broadcast here carries its angles only
-- so this lab can overlay and compare.
--
-- Port of src/client/labs/11-deterministic-rng/.
--
local sim = require 'playground.sim'
local app = require 'playground.app'
local Predict = require 'colyseus.predict.predict'

local kb = app.kb
local FAN_LEN = 40.0
local FAN_FADE_MS = 2400

local Lab = {
  id = "11-deterministic-rng",
  num = 11,
  title = "Deterministic Randomness",
  blurb = "Same seed both sides, nothing on the wire.",
  room_name = "lab-range",
}
Lab.__index = Lab

function Lab.new()
  return setmetatable({
    aim_x = 50, aim_y = 20,
    pending_fire = false,
    cheat = false,
    cheat_rng = sim.mulberry32(0x1234567),
    fans = {},
    max_divergence = nil,
  }, Lab)
end

function Lab:ready(room)
  return room.state.players ~= nil and room.state.players[room.session_id] ~= nil
     and room.state.bots ~= nil and room.state.bots["bot1"] ~= nil
end

function Lab:mount(context, room)
  self.room = room
  self.sid = room.session_id
  self.me = room.state.players[self.sid]
  self.bot = room.state.bots["bot1"]
  if self.me == nil or self.bot == nil then return false end

  self.predict = Predict.for_room(room)
  -- Bots ride the lerp timeline — the one the server rewinds to.
  self.predict:attach_all("bots", { "x", "y" },
    { mode = "lerp", delay = sim.REMOTE_INTERP_MS })
  self.predict:attach_all("players", { "x", "y" }, { mode = "damped" })

  room:on_message("spread", function(m) self:on_spread(m) end)

  self.input = room:input({ allow_rewind = function(data) return data.fire end })
  self.cmd = self.input.data
  self:build()
  return true
end

function Lab:build()
  self.recon = self.predict:reconciler(self.me, {
    input = self.input,
    fields = { "x", "y", "vx", "vy" },
    smoothing = 15,
    step = function(ctx, p, inp) sim.step_entity(p, inp.moveX, inp.moveY, ctx.dt) end,
  })
end

--- The client's half of the derivation — identical to the server's.
function Lab:client_fan(base_angle, seq, salt)
  if not self.cheat then return sim.spread_angles(base_angle, seq, salt) end
  -- The broken version: a local RNG the server cannot reproduce.
  local out = {}
  for i = 1, sim.PELLETS do
    out[i] = base_angle + (self.cheat_rng() - 0.5) * sim.SPREAD_RAD
  end
  return out
end

--- The server's fan for the same (seq, salt) — for overlay only.
function Lab:on_spread(m)
  if m.sid ~= self.sid or m.seq == nil then return end
  for _, f in ipairs(self.fans) do
    if f.seq == m.seq then
      local worst = 0
      for i = 1, math.min(#m.angles, sim.PELLETS) do
        f.server[i] = m.angles[i]
        local d = math.abs(f.client[i] - f.server[i])
        if d > worst then worst = d end
      end
      f.hits = m.hits or 0
      f.answered = true
      self.max_divergence = worst
      return
    end
  end
end

function Lab:fire() self.pending_fire = true end

function Lab:answered_fans()
  local n = 0
  for _, f in ipairs(self.fans) do if f.answered then n = n + 1 end end
  return n
end

function Lab:frame(context, now, dt_ms)
  if kb.autopilot then
    self.aim_x = self.predict:value(self.bot, "x")
    self.aim_y = self.predict:value(self.bot, "y")
  end
  if kb.key("space") then self.pending_fire = true end
  if kb.key("x") then self.cheat = not self.cheat end

  local steps = self.predict:tick(now)
  for _ = 1, steps do
    self.cmd.moveX = kb.move_x()
    self.cmd.moveY = kb.move_y()
    self.cmd.aimX = self.aim_x
    self.cmd.aimY = self.aim_y
    self.cmd.fire = self.pending_fire
    self.cmd.spread = true
    local seq = self.input:send()
    if self.pending_fire and seq and seq > 0 then
      local px, py = self.recon.state.x, self.recon.state.y
      local base = math.atan2(self.aim_y - py, self.aim_x - px)
      table.insert(self.fans, {
        seq = seq, ox = px, oy = py, t = now, answered = false, hits = 0,
        client = self:client_fan(base, seq, self.room.state.salt),
        server = {},
      })
      if #self.fans > 4 then table.remove(self.fans, 1) end
      self.pending_fire = false
    end
  end
end

function Lab:render(gfx)
  local now = app.now_ms()

  self.room.state.players:each(function(p, key)
    if key ~= self.sid then
      gfx.square(self.predict:value(p, "x"), self.predict:value(p, "y"),
        sim.PLAYER_HALF, gfx.hue(p.hue, 0.4))
    end
  end)

  local px, py = self.recon:value("x"), self.recon:value("y")
  gfx.square(px, py, sim.PLAYER_HALF, gfx.hue(self.me.hue))
  gfx.square_outline(px, py, sim.PLAYER_HALF, gfx.PALETTE.text, 1)

  local bx, by = self.predict:value(self.bot, "x"), self.predict:value(self.bot, "y")
  gfx.circle(bx, by, sim.BOT_RADIUS, gfx.a(gfx.PALETTE.blue, 0.25))
  gfx.circle_outline(bx, by, sim.BOT_RADIUS, gfx.PALETTE.blue)
  gfx.circle_outline(self.aim_x, self.aim_y, 0.9, gfx.a(gfx.PALETTE.text, 0.7))

  for i = #self.fans, 1, -1 do
    local f = self.fans[i]
    local age = now - f.t
    if age > FAN_FADE_MS then
      table.remove(self.fans, i)
    else
      local a = 1 - age / FAN_FADE_MS
      for p = 1, sim.PELLETS do
        local ang = f.client[p]
        gfx.line(f.ox, f.oy, f.ox + math.cos(ang) * FAN_LEN, f.oy + math.sin(ang) * FAN_LEN,
          gfx.a(gfx.PALETTE.warn, a * 0.75), 1.2)
      end
      if f.answered then
        for p = 1, sim.PELLETS do
          local ang = f.server[p]
          if ang then
            gfx.dashed(f.ox, f.oy, f.ox + math.cos(ang) * FAN_LEN,
              f.oy + math.sin(ang) * FAN_LEN, gfx.a(gfx.PALETTE.text, a * 0.9), 0.8)
          end
        end
      end
    end
  end
  if #self.fans > 0 then
    local last = self.fans[#self.fans]
    gfx.label(last.ox, last.oy, "amber = client — white = server",
      gfx.a(gfx.PALETTE.text, 0.5), 9, 14)
  end

  local last_hits
  for i = #self.fans, 1, -1 do
    if self.fans[i].answered then last_hits = self.fans[i].hits; break end
  end

  gfx.hud_section("TELEMETRY")
  gfx.hud_row("fan divergence (last shot)",
    self.max_divergence and string.format("%.5f rad", self.max_divergence) or "--",
    self.max_divergence == nil and gfx.PALETTE.text_faint
      or (self.max_divergence < 1e-6 and gfx.PALETTE.good or gfx.PALETTE.bad))
  gfx.hud_row("pellets hit (last shot)",
    last_hits and string.format("%d / %d", last_hits, sim.PELLETS) or "--",
    last_hits and gfx.PALETTE.text or gfx.PALETTE.text_faint)
  gfx.hud_row("room salt", tostring(self.room.state.salt), gfx.PALETTE.text)
  gfx.hud_row("shots (authoritative)", tostring(self.me.shots), gfx.PALETTE.text)

  gfx.hud_section("CONTROLS")
  gfx.hud_key("WASD", "drive")
  gfx.hud_key("mouse", "aim")
  gfx.hud_key("click / SPACE", "fire a 6-pellet fan")
  gfx.hud_key("X", self.cheat and "CHEATING with a local RNG" or "seeded from (seq, salt)")
  gfx.hud_note("Amber = the fan your client derived at the click. White dashes = " ..
    "the fan the server derived from the same (seq, salt). They should be " ..
    "identical to the pixel. Press X to swap in an unshared RNG and watch them " ..
    "disagree.")
end

function Lab:unmount() self.predict:dispose() end

function Lab:on_reconnect()
  self.me = self.room.state.players[self.sid]
  self.bot = self.room.state.bots["bot1"]
  if self.me == nil then return end
  self:build()
end

return Lab

File: clients/godot-app/scripts/labs/Lab11.cs
using System.Collections.Generic;
using System.Threading.Tasks;
using Colyseus;
using Colyseus.Predict;
using Godot;
using Lab = PredictProbe.LabSchema;

namespace Playground
{
    /// <summary>
    /// Lab 11 — Deterministic Randomness.
    ///
    /// The shotgun fan is "random", yet the client predicts every pellet EXACTLY
    /// — because the randomness is a pure function of data both sides already
    /// share:
    ///
    ///   seed = splitmix32(input seq ^ room salt)  →  mulberry32 pellet stream
    ///
    /// The seq is the engine's own input counter (Send() returns it; the server
    /// reads channel.consumedCount). The salt is synced room state. NOTHING
    /// about the pellets rides the wire — the server's broadcast here carries
    /// its angles only so this lab can overlay and compare.
    /// Port of labs/11-deterministic-rng/.
    /// </summary>
    public partial class Lab11 : LabBase<Lab.RangeState>
    {
        public override string Id => "11-deterministic-rng";
        public override int Num => 11;
        public override string Title => "Deterministic Randomness";
        public override string Blurb => "Same seed both sides, nothing on the wire.";

        private const double FanLen = 40.0;
        private const int MaxFans = 4;
        private const double FanFadeMs = 2400;

        /// <summary>One fired fan: what we derived, and what the server did.</summary>
        private class Fan
        {
            public int Seq;
            public double Ox, Oy;
            public readonly double[] Client = new double[Sim.Pellets];
            public readonly double[] Server = new double[Sim.Pellets];
            public bool Answered;
            public int Hits;
            public double T;
        }

        private Predict _predict;
        private Reconciler<Lab.RangePlayer, Lab.RangeInput> _recon;
        private Lab.RangeInput _cmd;
        private InputHandle _input;
        private Lab.RangePlayer _me;
        private Lab.Bot _bot;
        private string _sid;

        private double _aimX = 50, _aimY = 20;
        private bool _pendingFire, _cheat;
        private Sim.Rng _cheatRng = new Sim.Rng(0x1234567u);

        private readonly List<Fan> _fans = new List<Fan>();
        private double _maxDivergence;
        private bool _hasDivergence;

        public override async Task<bool> Mount(App app)
        {
            Room = await Shell.JoinLab<Lab.RangeState>(app, "lab-range",
                r => r.State.players != null && r.State.players.ContainsKey(r.SessionId)
                     && r.State.bots != null && r.State.bots.ContainsKey("bot1"));
            _sid = Room.SessionId;
            if (Room.State.players == null || !Room.State.players.TryGetValue(_sid, out _me)) return false;
            if (Room.State.bots == null || !Room.State.bots.TryGetValue("bot1", out _bot)) return false;

            _predict = Predict.Get(Room);
            // Bots ride the lerp timeline — the one the server rewinds to.
            _predict.AttachAll("bots", new AttachConfig {
                ["x"] = new PredictFieldOptions { Mode = PredictMode.Lerp, Delay = Sim.RemoteInterpMs },
                ["y"] = new PredictFieldOptions { Mode = PredictMode.Lerp, Delay = Sim.RemoteInterpMs },
            });
            _predict.AttachAll("players", new AttachConfig {
                ["x"] = PredictMode.Damped, ["y"] = PredictMode.Damped,
            });

            Room.OnMessage<Dictionary<string, object>>("spread", OnSpread);

            _cmd = new Lab.RangeInput();
            _input = Room.Input(_cmd, new InputOptions
            {
                AllowRewind = data => ((Lab.RangeInput)data).fire,
            });
            Build();
            return true;
        }

        private void Build()
        {
            _recon = _predict.Reconciler(_me, new ReconcilerOptions<Lab.RangePlayer, Lab.RangeInput>
            {
                Input = _input,
                Fields = new[] { "x", "y", "vx", "vy" },
                SmoothMs = 65,
                Step = (ctx, p, cmd) =>
                {
                    var e = new Sim.Entity { x = p.x, y = p.y, vx = p.vx, vy = p.vy };
                    Sim.StepEntity(ref e, cmd.moveX, cmd.moveY, ctx.Dt);
                    p.x = (float)e.x; p.y = (float)e.y; p.vx = (float)e.vx; p.vy = (float)e.vy;
                },
            });
        }

        /// <summary>The client's half of the derivation — identical to the server's.</summary>
        private void ClientFan(double baseAngle, int seq, uint salt, double[] outAngles)
        {
            if (!_cheat) { Sim.SpreadAngles(baseAngle, seq, salt, outAngles); return; }
            // The broken version: a local RNG the server cannot reproduce.
            for (int i = 0; i < Sim.Pellets; i++)
                outAngles[i] = baseAngle + (_cheatRng.Next() - 0.5) * Sim.SpreadRad;
        }

        /// <summary>The server's fan for the same (seq, salt) — for overlay only.</summary>
        private void OnSpread(Dictionary<string, object> msg)
        {
            if (!msg.TryGetValue("sid", out var sid) || (string)sid != _sid) return;
            if (!msg.TryGetValue("seq", out var seqRaw)) return;
            int seq = System.Convert.ToInt32(seqRaw);

            Fan fan = _fans.Find(f => f.Seq == seq);
            if (fan == null) return;
            if (!msg.TryGetValue("angles", out var anglesRaw) || !(anglesRaw is IList<object> angles)) return;

            double worst = 0;
            for (int i = 0; i < angles.Count && i < Sim.Pellets; i++)
            {
                fan.Server[i] = System.Convert.ToDouble(angles[i]);
                double d = System.Math.Abs(fan.Client[i] - fan.Server[i]);
                if (d > worst) worst = d;
            }
            if (msg.TryGetValue("hits", out var hits)) fan.Hits = System.Convert.ToInt32(hits);
            fan.Answered = true;
            _maxDivergence = worst;
            _hasDivergence = true;
        }

        public override void Frame(App app, double now, double dtMs)
        {
            var v = app.View;
            Vector2 mouse = Kb.MousePos;
            bool overStage = mouse.X < app.Stage.End.X;
            if (Kb.Autopilot)
            {
                _aimX = _predict.Value(_bot, "x");
                _aimY = _predict.Value(_bot, "y");
            }
            else if (overStage)
            {
                _aimX = v.WX(mouse.X);
                _aimY = v.WY(mouse.Y);
            }
            if (overStage && Kb.MouseDown()) _pendingFire = true;
            if (Kb.Key(Key.Space)) _pendingFire = true;
            if (Kb.Key(Key.X)) _cheat = !_cheat;

            int steps = _predict.Tick(now);
            for (int i = 0; i < steps; i++)
            {
                _cmd.moveX = (sbyte)Kb.MoveX();
                _cmd.moveY = (sbyte)Kb.MoveY();
                _cmd.aimX = (float)_aimX;
                _cmd.aimY = (float)_aimY;
                _cmd.fire = _pendingFire;
                _cmd.spread = true;
                int seq = _input.Send();
                if (!_pendingFire || seq <= 0) continue;

                double px = _recon.State.x, py = _recon.State.y;
                var fan = new Fan
                {
                    Seq = seq, Ox = px, Oy = py, T = now,
                };
                ClientFan(System.Math.Atan2(_aimY - py, _aimX - px), seq, Room.State.salt, fan.Client);
                if (_fans.Count == MaxFans) _fans.RemoveAt(0);
                _fans.Add(fan);
                _pendingFire = false;
            }
        }

        public override void Render(App app)
        {
            var v = app.View;
            double now = RoomClock.GetNow();

            Room.State.players.ForEach((key, p) =>
            {
                if (key == _sid) return;
                Draw.Square(v, _predict.Value(p, "x"), _predict.Value(p, "y"),
                    Sim.PlayerHalf, Palette.Hue(p.hue, 0.4f));
            });

            double px = _recon.Value("x"), py = _recon.Value("y");
            Draw.Square(v, px, py, Sim.PlayerHalf, Palette.Hue(_me.hue));
            Draw.SquareOutline(v, px, py, Sim.PlayerHalf, Palette.Text);

            double bx = _predict.Value(_bot, "x"), by = _predict.Value(_bot, "y");
            Draw.Circle(v, bx, by, Sim.BotRadius, Palette.A(Palette.Blue, 0.25f));
            Draw.CircleOutline(v, bx, by, Sim.BotRadius, Palette.Blue);
            Draw.CircleOutline(v, _aimX, _aimY, 0.9, Palette.A(Palette.Text, 0.7f));

            for (int i = _fans.Count - 1; i >= 0; i--)
            {
                var f = _fans[i];
                double age = now - f.T;
                if (age > FanFadeMs) { _fans.RemoveAt(i); continue; }
                float a = (float)(1 - age / FanFadeMs);
                for (int p = 0; p < Sim.Pellets; p++)
                {
                    double ang = f.Client[p];
                    Draw.Line(new Vector2(v.SX(f.Ox), v.SY(f.Oy)),
                        new Vector2(v.SX(f.Ox + System.Math.Cos(ang) * FanLen),
                                    v.SY(f.Oy + System.Math.Sin(ang) * FanLen)),
                        Palette.A(Palette.Warn, a * 0.75f), 1.2f);
                }
                if (!f.Answered) continue;
                for (int p = 0; p < Sim.Pellets; p++)
                {
                    double ang = f.Server[p];
                    Draw.DashedWorld(v, f.Ox, f.Oy,
                        f.Ox + System.Math.Cos(ang) * FanLen, f.Oy + System.Math.Sin(ang) * FanLen,
                        Palette.A(Palette.Text, a * 0.9f), 0.8f);
                }
            }
            if (_fans.Count > 0)
            {
                var last = _fans[_fans.Count - 1];
                Draw.Label(v, last.Ox, last.Oy, "amber = client — white = server",
                    Palette.A(Palette.Text, 0.5f), 9, v.S(Sim.PlayerHalf) + 4);
            }

            int lastHits = -1;
            for (int i = _fans.Count - 1; i >= 0; i--)
                if (_fans[i].Answered) { lastHits = _fans[i].Hits; break; }

            var h = app.Hud;
            h.Section("TELEMETRY");
            h.Row("fan divergence (last shot)",
                _hasDivergence ? $"{_maxDivergence:F5} rad" : "--",
                !_hasDivergence ? Palette.TextFaint
                    : _maxDivergence < 1e-6 ? Palette.Good : Palette.Bad);
            h.Row("pellets hit (last shot)",
                lastHits >= 0 ? $"{lastHits} / {Sim.Pellets}" : "--",
                lastHits >= 0 ? Palette.Text : Palette.TextFaint);
            h.Row("room salt", Room.State.salt.ToString(), Palette.Text);
            h.Row("shots (authoritative)", _me.shots.ToString(), Palette.Text);

            h.Section("CONTROLS");
            h.Key("WASD", "drive");
            h.Key("mouse", "aim");
            h.Key("click / SPACE", "fire a 6-pellet fan");
            h.Key("X", _cheat ? "CHEATING with a local RNG" : "seeded from (seq, salt)");
            h.Note("Amber = the fan your client derived at the click. White dashes = the fan " +
                   "the server derived from the same (seq, salt). They should be identical to " +
                   "the pixel. Press X to swap in an unshared RNG and watch them disagree.");
        }

        /// <summary>Fire from the acceptance harness, exactly as a click would.</summary>
        public void Fire() => _pendingFire = true;
        public bool Cheat { get => _cheat; set => _cheat = value; }
        /// <summary>Worst client↔server pellet disagreement, in radians. NaN until answered.</summary>
        public double MaxDivergence => _hasDivergence ? _maxDivergence : double.NaN;
        public int AnsweredFans { get { int n = 0; foreach (var f in _fans) if (f.Answered) n++; return n; } }

        public override void Unmount() => _predict?.Dispose();

        public override void OnReconnect()
        {
            if (!Room.State.players.TryGetValue(_sid, out _me)) return;
            Room.State.bots.TryGetValue("bot1", out _bot);
            Build();
        }
    }
}
